File: python/atelier_bench/reports/part_usage.py
"""Part usage report."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ..contracts.report_contract import ReportRow, report_row
from ..support.abstract_report import AbstractReport

if TYPE_CHECKING:
    from ..dataset import Dataset


class PartUsageReport(AbstractReport):
    SLUG = "part-usage"

    def __init__(self) -> None:
        super().__init__(self.SLUG, "Part usage")

    def rows(self, data: "Dataset") -> list[ReportRow]:
        return [report_row(part.name, part.consumed_quantity()) for part in data.parts]

File: go/atelier/reports/part_usage.go
// Package reports carries the 24 concrete reports. Each one embeds a base
// from atelier/support, declares only Rows, and registers itself in an init()
// function; Slug, Title, CacheKey and TTLSeconds arrive by promotion.
package reports

import (
	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/dataset"
	"atelier.example/lane/atelier/support"
)

// Compile-time proof of satisfaction: this assertion is the ONLY textual
// link between the struct and the interface it satisfies.
var _ contracts.ReportContract = (*PartUsageReport)(nil)

// PartUsageReport reports part usage.
type PartUsageReport struct {
	support.BaseReport
}

// NewPartUsageReport builds the report with its slug and title.
func NewPartUsageReport() *PartUsageReport {
	return &PartUsageReport{BaseReport: support.NewBaseReport("part-usage", "Part usage")}
}

// Rows is the only contract method this type declares.
func (r *PartUsageReport) Rows(data *dataset.Dataset) []contracts.ReportRow {
	rows := make([]contracts.ReportRow, 0, len(data.Parts))
	for _, part := range data.Parts {
		rows = append(rows, contracts.Row(part.Name, float64(part.ConsumedQuantity())))
	}
	return rows
}

// init registers this report; the package-level registry is built by the 24
// init functions in this package, in file-name order.
func init() {
	Register(NewPartUsageReport())
}
